ta/pkcs11/src/entry.c

```diff
@@ -36,28 +37,54 @@ void TA_CloseSessionEntryPoint(void *session __unused)
 /*
  * Entry point for invocation command PKCS11_CMD_PING
  *
- * @ctrl - param memref[0]: unused as input argument
- * @in - param memref[1] or NULL: expected NULL
- * @out - param memref[2] or NULL: output data buffer for TA version info
- *
- * Return a PKCS11_CKR_* value
+ * Return a PKCS11_CKR_* value which is also loaded into the output param#0
  */
-static uint32_t entry_ping(TEE_Param *ctrl, TEE_Param *in, TEE_Param *out)
+static uint32_t entry_ping(uint32_t ptypes, TEE_Param *params)
 {
+	TEE_Param *ctrl = &params[0];
+	TEE_Param *out = &params[2];
 	const uint32_t ver[] = {
 		PKCS11_TA_VERSION_MAJOR,
 		PKCS11_TA_VERSION_MINOR,
 		PKCS11_TA_VERSION_PATCH,
 	};
+	uint32_t rv = PKCS11_CKR_ARGUMENTS_BAD;
 
-	/* Ctrl param does not carry input data: expected but not used here */
-	if (!ctrl || ctrl->memref.size != sizeof(uint32_t) || in ||
-	    !out || out->memref.size != sizeof(ver))
-		return PKCS11_BAD_PARAM;
+	if (ctrl->memref.size != sizeof(uint32_t) ||
```

Comment:
I'm sorry to disagree once more. I'm trying to make this as simple as possible, not trying to be a pain in the neck.

With this patch we're using the classic pattern where param types and param is supplied directly to the function which handles this particular command. But with this patch the handler function is still not completely in control of the parameters. This is a mistake in my opinion. I'd prefer the handler function (`entry_ping()` in this case) did all the checking, reading and writing of the parameters with no flexibility in the direction of the parameters (unless it's really needed by some function). See https://github.com/OP-TEE/optee_os/blob/master/ta/avb/entry.c for an example.

If you'd like to use common code to update `ctrl` add a helper function later when needed.

Reply:
I added these helper functions in latest proposal:
```
static uint32_t entry_ping(uint32_t ptypes, TEE_Param *params)
{
	if (!tee_param_0_no_input(ptypes, &params[0]) ||
	    !tee_param_is_none(ptypes, 1) ||
	    !tee_param_is_output_memref(ptypes, 2) ||
	    !tee_param_is_none(ptypes, 3) ||
	    params[2].memref.size != sizeof(ver))
		return PKCS11_CKR_ARGUMENTS_BAD;

	(...)
}
```

But I left to the main handler `TA_InvokeCommandEntryPoint()` to write the status into the output memeref param#0, because this function shall handle the TEE effective return value for `TEE_ERROR_SHORT_BUFFER` case.

Reply:
I still think that the parameter handling in AVB is simpler. But it's you writing the code so if this is simpler for you go ahead.

Reply:
In the end, I think you're right... These `tee_param_is_*()` don't really clarify the expected parameters. I push an update.